### DISCUSSION TITLE
Adobe Air future

### DISCUSSION THREAD

**Discussion by ghost** [airsdk/Adobe-Runtime-Support#1077] [General]
As we all know, Adobe Air is not used as much as Unity or Haxe or Unreal for game development.

The ActionScript3.0 community and Starling is also really small right now, very different from 10 years ago.

I am afraid in the future will no longer exist anymore or its support will be canceled. I think a lot of us has the same feeling.

I would like to know, we wold like to know what plans does Harman has for Adobe Air in the future.

Will Adobe Air starts including additional support for developing ANE's or creating ports to, for example, PlayStation 5 content?

Will Adobe Air just focus on stand a stable version for desktops and mobile and that's it? I really want to now. I really want to now if I can safely stay here for 10 years more or I should change to another program?

Don't think bad about me, I fucking love Adobe and Air, ActionScript3.0 and coding there. I never felt that comfortable. I just want to know.

**Comment by htmiel** [comment]
From what I've seen so far, Harman seem committed to continue to support and also improve AIR; I do not worry about having to switch to another platform anytime soon (at least considering Windows, iOS and Android). You can take a look at this from last year: https://github.com/airsdk/Adobe-Runtime-Support/issues/385

**Comment by ajwfrost** [comment]
Yes we're good for many more years.... 10 is a fair way into the future but I'm still optimistic about that!

Curious as to whether PlayStation 5 is a platform that there would be a lot of demand for; we do have to make semi-commercial decisions about things like that i.e. would it be something more people would then join a paying license tier in order to get access to.

Mostly, the features and platforms will be driven by the demands of the community within those commercial boundaries, so we don't have unlimited resources available but we would like to move with the times somewhat and improve various aspects/features and platform support - maybe dropping some of the old and less used platforms (particularly where this becomes impractical, e.g. the support of 32-bit iOS apps which isn't possible if we can't run a 32-bit tool on a Mac..).

Hope that helps..

**Reply by CoachNair** [reply]
With Adobe giving in to Steve Jobs vested  interest of doing away with Flash 
to many of the brilliants AS3 developers who rocked action scripting – fading away & moving on to other platforms…
to a host of fantastic third party tools  quite unlikely to be updated….  ( with thankfully Starling – distriqt – myflashlabs etc.. being the exceptions)
It is just natural for the Amyfoosh in many of us - to at all times keep worrying about the future of Adobe Air.   

At the moment most organizations would be skeptical of supporting any application developed on AIR platform,
not  because they doubt AIR’s capability - but majorly because they don’t hear any big sound being made about it – 
& till that happens – I am sure many of them will prefer to stay safe &  away.

So while your reply definitely is reassuring  Andrew….
My opinion is that the community will need to align together to bring in the future wave for AIR.
Of course ADOBE-HARMAN – contract clauses may well dictate a lot of future initiatives…
BUT IF in the coming times – apart from the dedicated effort to develop new releases…
We could CROWD FUND  a robust IDE on the lines of FLASHDEVELOP ( maybe extend the same? )
reach out to the amazing developers to contribute- extend on available libraries of Stage 3D / Dynamics / AR etc etc  & help them monetize their efforts
develop a website strongly promoting the strength & reach of AIR – motivate seasoned developers to return – showcase AIR apps & basically create a strong BUZZ 

 - THIS is the need of the hour – now that we have come this far… few more steps & we should be there. 
I am sure you will find strong support from the community if HARMAN spearheads it.

**Reply by ajwfrost** [reply]
@CoachNair  very well said :-)

The IDE is a very good point; FlashDevelop is great but Windows-only. I kind of like the concept of Moonshine and think that this has a lot of potential for becoming the go-to tool specifically for AIR, although of course VS Code with the extension from Josh is also a good way of doing things (not so sure about the debugging support there but it may just be me not setting it up right..)

I think we're moving (at last, albeit slowly) in the right direction with the "AIR Package Manager" tool that Distriqt are spearheading, and there's plans afoot for a developer-based website which can be a little more dynamic (and dare I say useful) than the Harman website where we're constrained by our corporate marketing standards etc. With people writing articles recently e.g. about Android App Bundles, it may be good to collate them into a single location..

But also the idea of bringing new folk into the fold is one we had been discussing internally again recently. There were a couple of thoughts:
a) should we look to switch from ActionScript to JavaScript (or TypeScript) as the language used within AIR? This would of course be a major shift but we could probably keep the APIs 99% the same. I'm in two minds here as I much prefer AS3 to either of JavaScript or TypeScript! but supporting JavaScript would open it up so much... although the one technical challenge with that would be how to then cope with iOS where we're not allowed to just embed a JavaScript VM...
b) to help spread the word and get people familiar with AS3 and AIR, we could look to create a schools/colleges program, providing more tailored contents/tutorials etc with people just starting out with animating and programming. A lot of the basic queries about ActionScript seem to come from people who are set assignments with Adobe Animate and are trying to make things work in Flash..!

Anyway .. digressing, but yes I'm very much hoping we can increase this community! whilst also being aware that mostly what we still need to focus on is eliminating all the pain and workarounds that happen when using the IDEs and our SDK versions with the latest Android/iOS/macOS requirements...

**Reply by GeneralVimes** [reply]
> what we still need to focus on is eliminating all the pain and workarounds that happen when using the IDEs and our SDK versions with the latest Android/iOS/macOS requirements...

Indeed, the ever-changing platforms' requirement is something which bothers the developers the most. And we are so much thankful for the hard work Harman is doing for us. With x64 and .aab covered, what is the current most urgent need? iOS 15? Hopefully, after that someone like Google or Apple won't come up with new requirement, like "OK, developers, from yesterday you all have to submit your app as .qwerty file, and to make this .querty file you have to use only our own development tool". Or will they? :D

OK, so, after iOS15 support will the AIR team have more time to look around at the growth possibilities and new directions, please, correct me if I'm wrong.

As for the IDE, I'm using Flash Develop from Windows. It works great. Currently I'm struggling to understand how to set up environment on Linux machine to be able to produce a Linux build of my game. And here comes the first idea: knowledge base for AIR development. Like, sample projects for the popular IDEs and the platform targets. List of HowTo's. Code snippets. It can be done on a wiki engine and hosted somewhere (maybe this can be a community-driven initiative).

Second, there should be "Made with AIR" showcase. There are now Starling games showcase: https://wiki.starling-framework.org/games/start
Facebook group: https://www.facebook.com/groups/AdobeAirDevelopers
Showcase thread in https://www.reddit.com/r/as3/

Maybe there should be something centralized, curated by the company? Or maybe a twitter hashtag #madewithair should be launched?

**Reply by GeneralVimes** [reply]
> should we look to switch from ActionScript to JavaScript (or TypeScript) as the language used within AIR?

I strongly oppose switching to Javascript from AS3. Just my personal opinion, but I was making the same game for web browsers in JS and for mobile and PC in AS3, and AS3 development was going so much more pleasantly. (BTW, both versions of the game appeared to be good and well accepted by the audience, just, working in AS3 is a much more pleasant process).

If you want to expand the development community with those who know js, can you let the developers select a language? Like, in Unity, it's possible to code in C# or JS.

**Reply by GeneralVimes** [reply]
> Curious as to whether PlayStation 5 is a platform that there would be a lot of demand for; we do have to make semi-commercial decisions about things like that i.e. would it be something more people would then join a paying license tier in order to get access to.

As for the additional platforms support, I'd vote (both with voice and with money) for HTML5 platform. Can a possibility for the developers to write games in AS3 and convert them into HTML5 be added?

**Reply by GeneralVimes** [reply]
> ope with iOS where we're not allowed to just embed a JavaScript VM

By the way, this reminds me... It's possible to build .ipa on Windows. Rumors say that it's so thanks some agreement between Apple and Adobe. What do you think, can this be named as on of the AIR's unique strength compared to other game development engines and promoted? Like "Do you want to publish you game in App Store? No need to buy a Mac, use AIR!"

But this feature can be marketed only if there is an deal which Apple recognizes. Is there a chance that someone in Apple looks at this and says "Oh, no! Someone can build .ipa from Windows! No way! We must forbid this at once!" and will issue new requirement.

**Reply by GeneralVimes** [reply]
> At the moment most organizations would be skeptical of supporting any application developed on AIR platform

You know, for some time I thought that Google Play itself has prejudice against games made with AIR. So, as soon as the possibility to remove air. tag from the application descriptor appeared, I used it. And then I found out that Rusty Lake (https://play.google.com/store/apps/details?id=air.com.RustyLake.CubeEscapeParadox) was selected as the Best game at Google Play. So, at least there AIR games are treated the same way like the games made with different tools.

**Reply by CoachNair** [reply]
Thanks for the reply Andrew
Please find below my humble thoughts.  

ON IDE

Well I am definitely trying to reach out to one of the developer of FD - Philippe Elsass for a comeback  
Moonshine – have not checked but seems to have a similar UI + is alive-  which is what matters I suppose
– Will need to check the features as compared to FD.
But then again too many different ways of achieving the same goal – each with varying results can also 
make existing developers weary  &  newcomers feel lost ,
[ Moonshine  /  Josh-VS Code  /   IntelliJ IDEA  /   FDT (Not sure support exist) ] & probably others….


ON ELIMINATING ALL THE PAIN IN USING IDE’S….

YES  Which is absolutely true & very much required. 
But if we do not push now  – AIR will be in grave danger of getting phased out eventually– given the industry reluctance to accept AIR.
Currently even the seasoned developers are happy to have their existing apps kept alive with regular SDK updates rather than build new ones.
Feel we need to be focused like UNITY if we need to survive – fightback & establish AIR
- minus the GUI of course. Don’t need to make it clunky – Am just referring to their aggressive single track approach.
One dedicated AIR  IDE for Windows & Mac – as simple as FLASHDEVELOP in terms of installation & usage
with curated & tested ANE’s  / SWC’s integrated seamlessly with every upgrade ( especially of developers who stay committed - with varying pricing in accordance )
Given the right encouragement –to the right team -  this can come to be – given that monetizing it would not be such a big challenge.



ON ACTIONSCRIPT TO JAVASCRIPT

Apart from the disadvantages that you have yourself pointed out – the monumental language shift + iOS issues etc..
would it not also negate the very objective of motivating AS3 enthusiast to believe in AS3 & keep developing? 
Making developers wonder why bother with AIR at all – if it’s to be about JavaScript or typescript. 
Think that would be the final nail in the coffin for AS3.

Just asking - Would it not be possible instead to have a JS- AS3 AIR interpreter in the form of an ANE-SWC  to do the trick ?
Johnbrandle  with his JS -AS3 compiler seems to have found a solution to translate AS3 to JS - 
https://github.com/johnbrandle/as3-js
Maybe we could find a way to embed it too as an ANE  – So that language remains AS3.
AIR would be a potent force to reckon with if that could happen.
Also  – once we achieve the basics – which is the industry accepting AIR -
then the future can experiment with many promising possibilities with HARMAN’s expertise in the subject.
AS4 which was supposed to resurrect Flash but never got going could possibly become a work in progress  


ON WEBSITE & OTHER INITIATIVES

YES THANKS  –This is very interesting - 
Which again raises the point that if HARMAN  could partner with the platform that looks the most promising in terms of an AIR IDE
Then an awesome website homepage like   https://www.unrealengine.com/en-US/   ( with what is possible in AIR of course)
& an  integrated download link  [ IDE + SDK ]
– for Users to  Download – Install – Get started  - would be the perfect fillip we all need so badly
to put to rest all fears on the future of AIR.
This if supported with AIR apps video gallery -  informative links – tutorials – information – FAQ – all very professionally designed & managed
+ promoted across social media & other digital platforms - it could very well be a great booster at the right time.

The challenge I feel is not in developing the solution – It’s already there in various forms.
– But bringing all stake holders together & closing it would be key here I do understand.
But who other than HARMAN is better equipped for this.

Tapping outstanding talents like Michael – Daniel – Josh – Alexander Zadorozhny, 
Rob Bateman,  Philippe ,  Luca Deltodesco, Trevor McCauley, Grant Skinner……..
& host of so many others - who we could try & invite to contribute to the website with their articles & valuable inputs.
Am sure there would be many AS3 enthusiast – volunteers willing to support you in building this website. 
In adherence to your rules & regulations of course 


Thanks again

**Reply by SUSDEV** [reply]
A lot of interesting thoughts in this, I wanted to share my own opinions:

- What about starting with an online zoom conference? Probably easy to set up and would allow us to see the faces of other AIR developers and their struggles.
- I use FDT and the team at Interactive Pioneers are magnificent. I know there aren't changes on the web site, but they are very committed to AIR and I would definitely encourage other Flash/AIR developers to consider FDT. It is a great tool (extension on Eclipse). I have reached out to the team a few times for assistance and they are super.
- I would be hugely supportive of more community efforts, because it is very difficult for us AIR developers to find knowledgeable enough  people who can assist -- and I mean assist by paying, not just free stuff. Maybe we can have an area in which people can list their skills/interests/tools, like a directory. For example, I am a 1-person shop (regarding AIR development) but I have an oversized influence on my market, and my customer base is growing well. Since my apps came out in 2012, and I am not very technically savvy with the internals of the publishing process, I desperately need a community that can help me get my app into packageable forms for iOS, Android, and Mac (I've got Windows easily covered).
- Re: JS, I'm not sure what really would be accomplished here, considering the big effort it would take to do it.
- Would love if I could agree with @GeneralVimes about ipa from Windows only, but just having put together my iOS app, I hate to say one still needs a Mac for things like storyboard and submission process. Ugh, the notarization process is terrible.

Andrew, you and your team are superstars, I am (I'm sure like others here) are so fortunate you guys stepped in!

**Reply by marchbold** [reply]
I've been thinking a lot about this lately too and the directly AIR should take with regards IDEs. I think Harman need to be very careful with the options they are creating with ADT and that there is a simple way to integrate this into an IDE, particularly on a per project basis, rather than the sdk specific configuration. 

We use IntelliJ regularly and it's good to note that their plugin in open source (https://github.com/JetBrains/intellij-plugins/tree/master/flex), not the IDE but at least we could suggest updates easily there.

Also I highly rate VSCode and Josh has done great work with the extension there. I would suggest if we are going to get involved in one as a community that VSCode is the best place for us to focus. Super easy to update and build and a fast and very popular editor. Currently I see this as the easiest integration for `apm` as well, and can see some really great package manager integrations buildable in VSCode. 

I'm a little confused by the talk of JS, I think that would just open up the question of why don't I just use ReactNative or .... AS3 is one of the best points of AIR and I would highly discourage the thought of moving to the nightmare of JS where you need TS for it to be useable. I would suggest focussing on a web output instead, or developing the AS language further. 

@GeneralVimes #madewithair is a great idea too, however probably best driven by the community? We tried doing one on airnativeextensions, but I believe the stories around this are best driven by developers in a social platform.

**Reply by ghost** [reply]
Oh, boy. ~ 

I think I made a really interesting discussion point here, right? Ha ha.

I mean, I read all coments and I can just agree and disagree with some things here.

I don't think switching the entire language will be a solution. Being able to create a port as easy as I can say, "Alright, I want an .aab file" or "Alright, let's create an .exe or an .apk". The same with As3 > HTML5 is a solid and better solution for the future.

You know, Air is like Java: "Code one, export to anywhere". Adding more platforms like so is a lot smarter than just deleting as3 from the entire world.

Obviously, we're now saying: "Harman, tomorrow we want Switch, PS5 and VR games made in Air. Do it".

Oh, no no. ~

I know, we know you have some contracts and not unlimited money as Unity or Unreal developers has due to community length.

We're a small island on the ocean fighting againts america and china together.

I know you want the best for us, the users, and we can provide a lot of ideas, but you have the last word. And... That can be a really hard spot. I mean, a single mistake of yours can destroy the entire community forever. Scary. ~

Anyways...

VS Code is the top tier tool for coding right know. I don't think a new IDE for Air would be a solution. Because, you know, you have some IDE's to choose from and they have some plugins for As3 and they are very decent. This is not obviously a NEED right now.

My thoughts are:

— As3 > HTML5 support (as3 to HTML).
— A website with Air tutorials and a nice page to read documentation and download updates.

Both of them are for me the real thing to focus on.

**Comment by johnbrandle** [comment]
AS3-JS could potentially help with such an effort (BTW, it also has pretty accurate Flash API support). One idea to keep the language fresh might be creating a new AS3 specification (AS3.1).  Consisting of a subset of the AS3 specification (excluding particularly tricky, rarely used, features like named closures in conditionals), plus select features from es6 (compiled Babel-like to standard AS3, then to JavaScript). This language would have the best of both worlds, and would be compatible with both platforms.

Personally, I don't think the ActionScript language is the core issue. Many niche languages exist and are thriving. I believe the core issue is that AIR doesn't provide anything one cannot find elsewhere. The Flash ecosystem used to provide something none of the competition had: Flash Professional. This tool allowed developers, designers, and animators to work collaboratively, creating beautiful interfaces with unrivaled speed. However, Adobe made a mistake, rather than focusing on what made their platform unique, they began focusing almost entirely on performance. First came AS3 (alienating those with weaker dev backgrounds), then Stage3D (driving out any non-technical person still hanging on), and lastly came a slew of frameworks and libraries designed to work apart from Flash Professional. Ignoring Flash Professional, the ecosystem wasn't offering anything one couldn't find elsewhere. So, when Jobs finally did his thing, the majority didn't put up much of a fight (for good reason). If AIR is to succeed, Harman needs to bring back that tool in some form or another (with responsive and retina support). Similar tools have started to crop up to fill the void (Sketch, Figma, Zeplin, etc.), but none compare to Flash Professional. If Harman created an online equivalent (with HTML5 output and AIR support), it would provide a viable justification for moving to the ecosystem.

**Reply by marchbold** [reply]
Flash Professional was renamed to Animate and still functions in a similar way to develop AIR applications and animation assets. It's definitely still one of the tools we use regularly to create and edit assets for our AIR applications. 

However it has never been a good development environment and has always sat alongside tools such as FlashDevelop / FlashBuilder / FDT / IntelliJ.  Anyone serious about app development requires a more complete development environment which is the important point going forward.

**Reply by johnbrandle** [reply]
This "anyone serious about app development" mentality very much led to Flash going down the tubes. Was there anything serious about AS1? AS2 even? If the target audience is the truly serious, they were irrevocably lost to tools such as Xamarin and Unity quite some time ago. Do you think a "more complete development environment" is going to bring them back? No. Besides a love for AS3, why should anyone choose AIR over a myriad of technologies that accomplish the same thing (regardless of any dev environment improvements)? 

The real (non niche) target audience for this technology are those whom want to produce amazing results, quickly, simply, and cost effectively; those whom are not obsessed by that which has no practical significance, including feeling like a serious developer. The very same people whom made Flash what it was before Adobe took over (animators, designers, and amateur devs). That group was abandoned with the advent of Stage3D, and follow up technologies targeted solely to that "serious" crowd (most of whom were all to happy to jump ship for the next shiny thing). Essentially, Adobe lost sight of what made their offering unique (and I think its remaining users did too)...productivity.

Yes, the editor still exists in the form of Animate, but last time I checked (it was a fair time ago) it lacked native support for much needed things such as high res images and responsive UI. Additionally, now that real-time online collaboration is becoming the norm, I don't know that simply adding those features would be enough.

**Reply by CoachNair** [reply]
True - An online solution for HTML 5 output would definitely be a game changer & this with other dedicated AIR developments on Stage3D -Dynamics - AR – JS .. etc - could be the significant additions to justify the development of the language to be termed as AS3.1 or AS4.

Addressing the point mentioned here of the community drifting away - Adobe with its freely distributed SDK - possibly with the goal being to let flash dominate the marketplace - probably found itself at the receiving end - with Flash Builder losing competition to other free/ less costly / more robust - IDE's & the designer + developer community jumping at the opportunity to create beautiful outputs for free or nominal cost.

It still doesn’t explain though why  Adobe dint pursue AIR aggressively when the potential it had was so very evident… 
with the community surging ahead developing fantastic solutions & third party tools using the free SDK.
An integrated solution like Unity was well within their hands – but they seemed relieved to let Apple show the way – OUT.

Adobe did a LAST - LAME attempt to compete with UNITY – by projecting AIR as the GO-TO  tool to make 2D & 3D games.
BUT  this they claimed majorly on the strength of formidable third party solutions like Away3D & Starling.

I beg to differ though on your point that AS3 + Stage3d + numerous frameworks + libraries was the reason for Flash – AIR  downfall.
That Adobe chose to shift is the only reason why the community drifted – They were very much there all along till they
announced their decision to discontinue Flash on  web – without putting up a fight nor providing a solution for HTML 5 output.

Designers & Developers always CO-EXISTED  
I am saying this with absolute confidence because I have NO programming background whatsoever 
& many like me were actually overjoyed when something beautiful like AS1 – AS2 - AS3 came along with a host of other powerful libraries. 
There was never any technologically driven stress on a designer.
Flash Professional allowed them to explore their creativity - with developers breathing life into it with interactivity.
THAT was when many of us designers could actually explore the world of coding much to our delight.

My fear is something else 
Having dabbled with Macromedia Director & seeing it killed by Adobe after they acquired it …
the pattern seems sort of similar – Adobe went through the gestures of supporting it but let FLASH + AS 1 & 2 take over gradually.
Interactive presentations ( CD/DVD)  powered by Director -  was the very happening thing then &  bread & butter for my production house. 
We the very hardcore Director loyalist succumbed so easily to Flash – because Flash was a 2 in 1 solution. 
Interactive presentations ( CD/DVD)  + the WEB .
Now while ADOBE has given  HARMAN the task of picking up the gauntlet  & they are doing exceptional work…
If ADOBE’s objective is to slowly  phase it out in a natural way – they may very well achieve it. 
What killed Director  was the WEB & ironically the very WEB which propelled FLASH – stood in its way - after years of dominance.
So if HTML5 support does not happen soon the drifting way from AIR most probably will.
Unless of course something relatively simple like the apps being accessed on mobile & desktop - can just as easily be accessed on the web - With Wasm or any other means. 

Time will tell.

**Reply by GeneralVimes** [reply]
Just was about to write, that the most massive flow of new developer was connected with the fact that in Flash you could draw something - and at once make the entities you have just drawn interact. Maybe something new can be made these days... No need of all the Animate's juggernaut features: just the easy way for newcomers to create a small game using pictures they have just drawn.

**Reply by manishpatel0404** [reply]
@GeneralVimes - Isn't that actually an idea for a simple gaming app framework developed in AS3/AIR?

**Reply by johnbrandle** [reply]
> Addressing the point mentioned here of the community drifting away - Adobe with its freely distributed SDK - possibly with the goal being to let flash dominate the marketplace - probably found itself at the receiving end - with Flash Builder losing competition to other free/ less costly / more robust - IDE's & the designer + developer community jumping at the opportunity to create beautiful outputs for free or nominal cost.

Exactly, there were no great alternatives to Flash Professional, but there were plenty of great alternatives to Flash Builder. So, when they began adding features that were not fully supported by Flash Professional (3D to compete with Unity), or just focusing on performance (for mobile and their growing convert crowd), they pushed their base farther and farther away from the money maker. As AS3, Stage3D, Starling, PureMVC, Away3D, etc. began to take hold, more and more of their long-time base were forced back to the simpler HTML world, and newbies were just SOL. However, this exodus was mitigated by transplants coming in from other backgrounds, particularly Java. At Disney for instance, most of the Flash teams were composed of Java or C++ engineers, having no prior experience with Flash. So, of course, when Adobe came around asking, how would you like to see Flash improved...what do you think was said? Certainly nothing in relation to Flash Professional! They were riding the Flash money tidal wave, had no loyalty to the tool, and rather than adapting to Flash, they made Flash adapt to them.  

Without Flash Professional, there was nothing special to be found for those who came from those other backgrounds. Love AS3? Ha, most wanted something more akin to C#. Why use Flash for 3D when you could use Unity with its WYSIWYG GUI and more technical language? Why use Flash for mobile development if you're not using the IDE and all you care about is performance? Why use AS3 at all? We had a guy writing a Java to AS3 compiler...why...because he was still a Java engineer who missed Java. Essentially, Adobe unwittingly exchanged its loyal fan base (who relied on Flash Professional) with those who were only using Flash to make a quick buck and/or a name for themselves (who could care less about the tool, or anything Adobe for that matter).

> Adobe did a LAST - LAME attempt to compete with UNITY – by projecting AIR as the GO-TO tool to make 2D & 3D games.
BUT this they claimed majorly on the strength of formidable third party solutions like Away3D & Starling.

This was a huge mistake on Adobe’s part. People were excited of course, who wouldn’t be? But unless Adobe added full 3D support to their editor, there was no competing with Unity once the excitement passed. To state things a bit clearer, try imagining Unity being as popular as it is today without supporting their WYSIWYG Editor. That editor made them, and it made Flash as well. Adobe failed to realize that. 

> I beg to differ though on your point that AS3 + Stage3d + numerous frameworks + libraries was the reason for Flash – AIR downfall.
That Adobe chose to shift is the only reason why the community drifted – They were very much there all along till they
announced their decision to discontinue Flash on web – without putting up a fight nor providing a solution for HTML 5 output.

Yes, had Adobe prioritized HTML5 output early on, I believe the inevitable would have been delayed. However, without the Flash Professional workflow, they would not have been able to offer anything that could not be found elsewhere. I can think of four things Adobe needed to survive for the long term:

All features to be fully compatible with Flash Professional 
HTML5 output
Adding vital features such as high resolution images and responsive UI 
Putting more focus on productivity and collaboration rather than performance

> Just was about to write, that the most massive flow of new developer was connected with the fact that in Flash you could draw something - and at once make the entities you have just drawn interact. Maybe something new can be made these days... No need of all the Animate's juggernaut features: just the easy way for newcomers to create a small game using pictures they have just drawn.

Yep, this flow is what most people appear to have forgotten. I remember having FLAs being handed off to me, fully laid out and animated. All I had to do was group a few items, give names, and I could manipulate everything via code. Need to transition to another screen? Just go to the frame! Need to play an animation? Just Play! I could have even the most complicated screens up and running within a few hours. What does that process look like today? For one, you have to deal with HTML and CSS embedded in your code (looking at you React). YOU have to lay it all out, and YOU have to animate it. YOU have to update it whenever the designer makes changes. It’s all a bit masochistic and masturbatory. Would I do it if I needed the performance? Absolutely, but most apps just don’t need it...and devs are paying a heavy price for that psychological silver star.

BTW, towards the end, the Flash ecosystem was quickly beginning to resemble what the HTML ecosystem is today (a hot mess).

**Reply by CoachNair** [reply]
True John – On all aspects.

Now cut to present - Adobe’s out – Animate will never be the new Unity.
& HARMAN am sure will not get into a WYSIWYG GUI based solution in the foreseeable future.
So the question to ask ourselves I suppose is – What next?

As of today - we have AIR SDK + Third party solutions (IDE’s + Frameworks + ANE/SWC’s Libs etc.)
At present - We can build reasonably powerful Desktop & Mobile applications with these existing tools.
In future - In the absence of a WYSIWYG GUI based solution which as said is impractical to expect - 
we could possibly look forward to HARMAN for an integrated SDK + IDE (by partnering with the best in fray) + seamless integration of active third party solutions within the IDE.
Which if practically spelled out would be – AIR SDK + IDE (FDT or VS Code) + Starling + Distriqt / Flash Labs ANE’s & any other active libraries - frameworks which could make the integrated solution more potent.  HTML5 output – Stage 3D – AR etc - could be the future additions – IF HARMAN finds it worth their while to pursue.

From a marketplace point of view – which is what finally dictates survival - I can see only the below possibility for AIR – given the above tools & the fact that HARMAN is periodically releasing updates.

An Individual or team of people - who funds & develops independently without the backing of any big investor - will keep going as long as they can.  A very arduous journey of waiting for popularity of their idea to gain ground by word of mouth for monetization to happen. Unless they spend additional for aggressive promotions too. 

Given the uncertainty surrounding AIR I feel – NO INVESTOR would be really interested in financing AIR solutions & the same would go for ORGANISATIONS who build applications for internal use or B2B solutions.
I have personally experienced the absolute lack of interest for an idea within my circle – with them questioning my choice of platform.

So going forward I feel - while at an individual level we could all keep marching with HARMAN – the real question should be – How do we get the INVESTORS & ORGANISATIONS to believe in the practicality of AIR usage to develop powerful solutions which will not die on them in the future. 
That AIR has a passionate following & a definite potential is not going to make them come back.

Would an Integrated IDE mentioned on the lines above & a very strong promotional BUZZ of its relevance – power & FUTURE – on the lines of what Andrew suggested + more do the trick?

**Reply by megajogos** [reply]
For me the most needed is the HTML output and improving the ANR rate.

I think the HTML output would make BUZZ need to get AIR back on the market as there are thousands of flash games that could work again.

We really like to develop using AIR and the IDEs or AS3 is not a problem for us.

**Comment by KaminoRyo** [comment]
It's a very interesting discussion. 
We need to sort out "strengths and weaknesses". It is necessary to organize "improvement measures and cost performance".

[note]
I was developing educational software using Flash / Adobe AIR from 2014 to 2019.
Right now I'm completely using TypeScript and React.

strength:
- Support for 4 (5) major OS.(Win/mac/iOS/Android/Linux).
- There is a GUI tool (= Adobe Animate) that allows you to easily write animations. 
- We have an existing code base (although it can be a weakness).

weaknesses:
- Adobe AIR does not work directly on the web.
- Languages(=ActionScript3) and specifications are outdated.
- The range of skills of the developer is too wide, and it is difficult to gather opinions.
- (I think )Adobe Animate will throw away Adobe AIR around 2025.

improvement measures and cost performance:
- Web support(publish to HTML/JS/CSS or WASM)
A rich experience on the web (preferably easy to develop) is the last sanctuary. 
Destroyed by Apple 10 years ago, its value remains unwavering.
However, due to development difficulties, Unity has withdrawn, Cocos2d-x has left, and Phaser and Play Canvas isn't the main player.
It's tough for Adobe AIR to be the main player. 

- Renewal of language and specifications
- Text based IDE improvements
Many programmers are overjoyed. But, It's likely that the user base will shrink. Weaknesses disappear, but does not change to strengths.
The cost performance is also delicate. (Especially when changing the language or specifications).

Thank you for reading the long, poor English text.

**Reply by GeneralVimes** [reply]
Great analysis! Yes, I totally agree that the web support is the first needed step forward. 
I have a question about, how is AS3 outdated? I mean, I don't quite grasp it. I understand that languages like GWBASIC or Pascal are outdated, but how is AS33 outdated? What features does it lack?

**Reply by CoachNair** [reply]
Don't think it's the lack of features that makes AIR seem outdated... it's the lack of conviction of Institutions & developer community of AIR's future that made them drift away to alternative solutions.

**Reply by CoachNair** [reply]
... consequently rendering it outdated

**Reply by KaminoRyo** [reply]
I'm very impressed with TypeScript and React. It is also influenced by functional programming.
This is likely to disagree with game programmers. 

In a word, it is "all". ActionScript 3 is a language born in 2006(?), so it can't be helped. 

- The type system is too old.(Difficult to explain...)
- Not null-safety.
- All asynchronous processing is an event handler.
- Added every year in ECMAScript, no convenient built-in methods.

Frankly speaking, it is difficult to create and maintain "AS3++" or "AS4", and I think there is still hope for accepting TS. 
I'm not familiar with the compiler, Will it be TS → (tsc) → JS (+ map file?) → (air compiler) → AIR Binary as an approach?

**Reply by KaminoRyo** [reply]
I understand the importance of being kind to beginners.
I also know that complex type systems and architecture theory hinder it.
But I think the main thing left now is a lot of advanced apps and their maintainers that are too costly to bring to other languages. 
Also, I don't think there will be new beginners in the current AIR. 
Therefore, I think there is a return to using resources for programmer requests. 😔

Now we are in a chicken or egg situation.🐣🐓

**Reply by ajwfrost** [reply]
@KaminoRyo I'm curious as to what you mean by "the type system is too old" and "not null-safety". AS3 is strongly typed and throws null reference errors if you hit them, which is similar to Java and C++, C# etc. AS2 was not strongly typed and didn't throw errors if you tried dereferencing something that was null/undefined, which is more similar to JavaScript, and is a problem if you're trying to develop large applications. I'm not sure we want to go back to that sort of behaviour - which is the reason why they created TypeScript in the first place, to deal with JavaScript's weaknesses in that sort of area..!

Anyway .. this is a great thread, we will come back to it repeatedly I think to get some insights, but it's fairly clear we can't please everyone, sorry!

**Reply by radubirsan** [reply]
Flash  missed the open source train 10 years ago. 
Today there is a new train and it is decentralized ( if you don't understand I don't have time to explain it to you ).
All flash needs for a rebirth, is a White Paper. Do you want to board the train?

**Reply by KaminoRyo** [reply]
@ajwfrost Well, from that question, do you mean that you are not very familiar with programming languages in the 10s?
It will be long to explain, so I want you to touch the TypeScript.
Of course, "TypeScript has a state-of-the-art and perfect type system" is a joke, there is also many points that can be used as a reference for what to do with ActionScript from now on.

**Reply by ajwfrost** [reply]
@KaminoRyo I was wondering whether you wanted to move more to a type-less JavaScript style language! which I always find to be a pain as it's harder to find errors. But yes TypeScript would be more similar to what we're looking at - someone had suggested somewhere else about the syntax like "a = b?.c", we can see whether that's possible to add in the compiler and whether it needs runtime updates too...

Regarding the idea of actually allowing development in TypeScript: the simplest option from our perspective would be for someone to create a compiler that takes TypeScript and generates ActionScript Byte Code. But that's a bit beyond normal TypeScript transpilation capabilities I think.. so you could always look at taking JavaScript and converting it to ActionScript. Or going via LLVM. Any such tool could of course be created by anyone/open source etc but it's telling how many folk have started and then abandoned efforts for TypeScript to LLVM etc.. So maybe just theoretical at this stage! Obviously for us, the simplest option would be to include the v8 engine within AIR and just allow folk to use normal tools for TypeScript/JavaScript development, exposing the AIR APIs. But for two things: (a) garbage collection, and (b) the AS3 APIs, the implementation of both of which are very much tied up with the avmplus code.

Anyway. Interesting thoughts and I appreciate the inputs :-)

thanks
